### PR TITLE
Cloud Spanner: Adds tracing 

### PR DESCRIPTION
**Pull request by vkedia**
cc @tomayles @bogdandrutu

### CODE REVIEW

File: google-cloud-spanner/pom.xml [hunk 1]
```diff
@@ -182,5 +182,10 @@
       <classifier>testlib</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>io.opencensus</groupId>
+      <artifactId>opencensus-api</artifactId>
+      <version>0.6.0</version>
```
Comment (bogdandrutu) [inline google-cloud-spanner/pom.xml]: Can you depend on 0.9.1?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceUtil.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      endSpanWithFailure(span, (SpannerException) e); 
+    } else {
+      span.end(EndSpanOptions.builder()
+          .setStatus(Status.INTERNAL.withDescription(e.getMessage()))
+          .build());
+    }
+  }
+  
+  static void endSpanWithFailure(Span span, SpannerException e) {
+    span.end(EndSpanOptions.builder()
```
Comment (bogdandrutu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceUtil.java]: In 0.9.x opencensus-api you have an explicit setStatus which will help. Our API is backwards compatible so you can depend on 0.9.1 in your library and exclude the transitive dependency from the gRPC.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceUtil.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        .setStatus(getOpenCensusStatus(e.getErrorCode()).withDescription(e.getMessage()))
+        .build());
+  }
+  
+  private static Status getOpenCensusStatus(ErrorCode code) {
```
Comment (bogdandrutu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/TraceUtil.java]: We have a package for grpc status conversion (if you want to use that):
opencensus-contrib-grpc-util

https://github.com/census-instrumentation/opencensus-java/blob/master/contrib/grpc_util/src/main/java/io/opencensus/contrib/grpc/util/StatusConverter.java

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 4]
```diff
@@ -495,6 +504,31 @@ String extractHeader(String resource) {
     return elements.get((int) index);
   }
 
+  /**
+   * This is a one time setup for grpcz pages. This adds all of the methods to the Tracing
+   * environment required to show a consistent set of methods relating to Cloud Bigtable on the
+   * grpcz page.  If HBase artifacts are present, this will add tracing metadata for HBase methods.
+   */
+  private static void setupTracingConfig() {
```
Comment (bogdandrutu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: Note that this is needed for gRPC version < 1.8.0 (if you change to gRPC 1.8.0 and regenerate the gRPC stubs using gRPC-proto-1.8.0) this is no longer needed. So add a message to remove this.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 5]
```diff
@@ -29,47 +45,106 @@
...
   @Override
   public ReadContext singleUse() {
-    return pool.getReadSession().singleUse();
+    Span span = tracer.spanBuilder(READ_ONLY_TRANSACTION).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      return pool.getReadSession().singleUse();
+    } catch (RuntimeException e) {
+      TraceUtil.endSpanWithFailure(span, e);
+      throw e;
+    }
```
Comment (bogdandrutu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: You probably need to have a finally here as well. (more cases like this in the rest of the file).

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java [hunk 6]
```diff
@@ -216,15 +244,22 @@ public void cancelled(Context context) {
...
-              public com.google.spanner.v1.Session call() throws Exception {
-                return rpc.createSession(db.getName(), getOptions().getSessionLabels(), options);
-              }
-            });
-    return new SessionImpl(session.getName(), options);
+    Span span = tracer.spanBuilder(CREATE_SESSION).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      com.google.spanner.v1.Session session =
+          runWithRetries(
+              new Callable<com.google.spanner.v1.Session>() {
```
Comment (bogdandrutu) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java]: FYI: You can add a TODO to change to the new API that we add in 0.10.0 to wrap callable/runnable see:
https://github.com/census-instrumentation/opencensus-java/pull/778

File: google-cloud-spanner/pom.xml [hunk 7]
```diff
@@ -153,5 +159,25 @@
       <classifier>testlib</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>io.opencensus</groupId>
+      <artifactId>opencensus-api</artifactId>
+      <exclusions>
+        <exclusion>
```
Comment (vam-google) [inline google-cloud-spanner/pom.xml]: Why is this required?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 8]
```diff
@@ -480,6 +490,37 @@ public void start(Listener<RespT> responseListener, Metadata metadata) {
...
   }
 
+  /**
+   * This is a one time setup for grpcz pages. This adds all of the methods to the Tracing
+   * environment required to show a consistent set of methods relating to Cloud Bigtable on the
+   * grpcz page.  If HBase artifacts are present, this will add tracing metadata for HBase methods.
+   *
+   * TODO: Remove this when we depend on gRPC 1.8
+   */
+  private static void setupTracingConfig() {
```
Comment (vam-google) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: Should this be configured in Rpc class (looks like a job for the utils class)?

File: google-cloud-spanner/pom.xml (hunk 7 above)
Comment (vkedia) [reply google-cloud-spanner/pom.xml]: I believe I was running into some issue with errorprone version conflicts. Let me try removing it and compiling again.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 8 above)
Comment (vkedia) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: This will be removed pretty soon since grpc will do it automatically with 1.8. So I did not bother to put it in a separate class.

**Comment by vkedia** [comment]
@vam-google Thanks for the approval. I agree it does make the code less readable but this is a pretty important feature.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 8 above)
Comment (vkedia) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: This is the error I get
```
[WARNING] Rule 3: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.google.errorprone:error_prone_annotations:2.0.19 paths to dependency are:
+-com.google.cloud:google-cloud-spanner:0.32.1-beta-SNAPSHOT
  +-com.google.truth:truth:0.30
    +-com.google.errorprone:error_prone_annotations:2.0.19 (managed) <-- com.google.errorprone:error_prone_annotations:2.0.8
and
+-com.google.cloud:google-cloud-spanner:0.32.1-beta-SNAPSHOT
  +-com.google.guava:guava-testlib:20.0
    +-com.google.errorprone:error_prone_annotations:2.0.19 (managed) <-- com.google.errorprone:error_prone_annotations:2.0.12
and
+-com.google.cloud:google-cloud-spanner:0.32.1-beta-SNAPSHOT
  +-io.opencensus:opencensus-api:0.9.0 (managed) <-- io.opencensus:opencensus-api:0.6.0
    +-com.google.errorprone:error_prone_annotations:2.0.19
and
+-com.google.cloud:google-cloud-spanner:0.32.1-beta-SNAPSHOT
  +-io.opencensus:opencensus-contrib-grpc-util:0.9.0
    +-com.google.errorprone:error_prone_annotations:2.0.19 (managed) <-- com.google.errorprone:error_prone_annotations:2.1.2
and
+-com.google.cloud:google-cloud-spanner:0.32.1-beta-SNAPSHOT
  +-io.grpc:grpc-netty:1.7.0
    +-io.grpc:grpc-core:1.7.0
      +-com.google.errorprone:error_prone_annotations:2.0.19 (managed) <-- com.google.errorprone:error_prone_annotations:2.1.2
]
```
